src/Collections/Sequences/LittleEndianNatConversions.dfy

```diff
@@ -81,15 +81,35 @@ abstract module {:options "-functionSyntax:4"} LittleEndianNatConversions {
     requires |xs| % E() == 0
     ensures |ys| == |xs| / E()
   {
-    if |xs| == 0 then LemmaDivBasicsAuto(); []
+    if |xs| == 0 then
+      var ys := (LemmaDivBasicsAuto(); []);
+      assert |ys| == |xs| / E();
+      ys
+
     else
       LemmaModIsZero(|xs|, E());
       assert |xs| >= E();
-
       Small.LemmaSeqNatBound(xs[..E()]);
-      LemmaModSubMultiplesVanishAuto();
       LemmaDivMinusOne(|xs|, E());
-      [Small.ToNatRight(xs[..E()]) as Large.uint] + ToLarge(xs[E()..])
+      assert |xs[E()..]| % E() == 0 by {
+        LemmaModSubMultiplesVanishAuto();
+      }
+      var ys := ([Small.ToNatRight(xs[..E()]) as Large.uint] + ToLarge(xs[E()..]));
+      assert |ToLarge(xs[E()..])| == |xs[E()..]| / E();
+      assert |ys| == |xs| / E() by {
+        calc {
+          |ys|;
+          1 + |xs[E()..]| / E();
+          1 + (|xs| - E()) / E();
+          { DivMod.ModINL.LemmaFundamentalDivMod(|xs|, E()); }
```

Comment:
I don't understand why this rewrite is so complicated.

Couldn't you do:

```dafny
|ys|;
1 + |xs[E()..]| / E();
1 + (|xs| - E()) / E();
1 + |xs| / E() - E() / E();
1 + |xs| / E() - 1;
|xs| / E();
```

Reply:
Your proof works too. The difference is that with my version it takes 1.35M RU, and yours takes 2.16M RU.
The reason why it's so complicated is that I wrote a detailed proof that ought to reduce the resource count, and then gradually removed elements that decreased the resource count, until I arrived at a minimum.

Reply:
I think it's good to optimize for resource count, but it should remain easy to understand why the code is the way it is. If special tricks are required to achieve a certain resource count, they should be explained. Currently I can't understand why this proof is the way it is. I would have expected the one with fewer steps to use fewer resources.

Reply:
I don't have a full understanding why some terms reduce the resource count and some don't. Some changes might not be statistically significant (like the last effort), some are surely (like putting the lemma ensuring quantifiers in an if then condition)
The only thing I know is that this proof works in 4.0.0, 4.1.0, 4.2.0, 4.3.0 when before it would not, so I'd say we keep it as such, until one day we have a better understanding about proof elements that can predictably and reliably reduce the resource count.

Reply:
Wouldn't the simple version also work with all those versions? If so I'd prefer that.

Otherwise, could you put a comment in the code to explain how you got to that proof?

Reply:
I added a comment about the process.
Note that this file does not use nonlinearities, so it's a bit puzzling that your proof works on the first place.

Reply:
> Note that this file does not use nonlinearities, so it's a bit puzzling that your proof works on the first place.

Maybe because it doesn't require interpreting any arithmetic operations? It only uses `X / X == 1` and the distributive law.

Reply:
`a/y` is not the same as `a(1/y)`. `1/y == 0` when `y > 1`